[SYCL][SCLA] Add CodeGen capabilities for `sycl_ext_oneapi_private_alloca`

victor-eds opened the pull request:
The [`sycl_ext_oneapi_private_alloca`](https://github.com/intel/llvm/blob/56e9067ba69809fb6ea1fd4328456ca3a009f984/sycl/doc/extensions/experimental/sycl_ext_oneapi_private_alloca.asciidoc) adds new functions returning a pointer to a specialization constant length SYCL private memory allocation. This commit adds codegen support for these functions.

The `sycl::private_alloca` function is implemented as an alias to a new `__builtin_intel_sycl_alloca` builtin. This is needed to guarantee the call will lower to just an `alloca` instruction defining the private memory allocation.

This builtin lowers to a SYCL builtin call to `__builtin_sycl_unique_stable_id` and a call to a new `llvm.sycl.alloca` intrinsic. This intrinsic receives three arguments encoding the specialization constant used as array length, a type hint argument encoding the allocation element type and the required alignment.

Note the `sycl_ext_oneapi_private_alloca` extension defines two functions: `private_alloca` and `aligned_private_alloca`. This commit adds codegen support only for the first signature, but already prepares support for the aligned flavor by adding an argument encoding the memory allocation alignment to the `llvm.sycl.alloca` intrinsic.

The intrinsic is needed as generating an `alloca` instruction right away would lead to optimization passes converting the size argument, which can be of any integral type, and thus difficulting lowering to a single SPIR-V `OpVariable` later in the pipeline.

maksimsab reviewed (approved):
Tools

premanandrao commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -177,6 +177,28 @@ def err_intel_sycl_ptr_annotation_mismatch
 
+def err_intel_sycl_alloca_no_alias
+    : Error<"__builtin_intel_sycl_alloca cannot be used in source code. "
+            "Use the private_alloca alias instead.">;
```
```suggestion
            "Use the private_alloca alias instead">;
```

premanandrao commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -23454,6 +23456,85 @@ RValue CodeGenFunction::EmitIntelSYCLPtrAnnotationBuiltin(const CallExpr *E) {
+  }
+
+  // If no slot is provided, simply return allocation.
+  if (ReturnValue.isNull()) {
```
No need to brace enclose single statement blocks.

premanandrao commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -7328,6 +7338,98 @@ bool Sema::CheckIntelSYCLPtrAnnotationBuiltinFunctionCall(unsigned BuiltinID,
+  if (CheckSize(getASTContext(), CST)) {
+    TemplateArgument TA = CST->get(1);
+    QualType Ty = TA.getNonTypeTemplateArgumentType();
+    if (Ty.isNull()) {
+      Diag(Loc, diag::err_intel_sycl_alloca_no_size) << TA;
+    } else {
+      Diag(Loc, diag::err_intel_sycl_alloca_wrong_size) << TA << Ty;
+    }
```
Would prefer if this was not brace-enclosed, but up to you.

premanandrao commented on clang/test/CodeGenSYCL/builtin-alloca.cpp:
```diff
@@ -0,0 +1,86 @@
+// RUN: | FileCheck %s --check-prefix=CHECK-DEVICE
+// RUN: %clang_cc1 -fsycl-is-host   -triple x86_64-unknown-unknown -disable-llvm-passes -emit-llvm -o - %s -fcxx-exceptions\
+// RUN: | FileCheck %s --check-prefix=CHECK-HOST
+
```
Please add a comment here about what this test tests for.

premanandrao commented on clang/test/CodeGenSYCL/builtin-alloca.cpp:
```diff
@@ -0,0 +1,86 @@
+constexpr sycl::specialization_id<int> intSize(-1);
+constexpr sycl::specialization_id<unsigned short> shortSize(1);
+
+// COM: On the device, for each call, we should generate a chain of: 'call @sycl.alloca.<ty>' + ('addrspacecast') + 'store'.
```
```suggestion
// On the device, for each call, we should generate a chain of: 'call @sycl.alloca.<ty>' + ('addrspacecast') + 'store'.
```
Likewise for all comments below; don't need the COM: prefix.

premanandrao commented on clang/test/CodeGenSYCL/builtin-alloca.cpp:
```diff
@@ -0,0 +1,86 @@
+// CHECK-HOST-NEXT:    [[PTR2:%.*]] = alloca %"class.sycl::_V1::multi_ptr.1", align 8
+// CHECK-HOST-NEXT:    store ptr [[KH]], ptr [[KH_ADDR]], align 8
+// CHECK-HOST-NEXT:    [[TMP0:%.*]] = load ptr, ptr [[KH_ADDR]], align 8
+// CHECK-HOST-NEXT:    [[CALL:%.*]] = call ptr @_ZN4sycl3_V13ext6oneapi12experimental14private_allocaIdTnRDaL_ZL4sizeELNS0_6access9decoratedE1EEENS0_9multi_ptrIT_LNS6_13address_spaceE0EXT1_EEERNS0_14kernel_handlerE(ptr noundef nonnull align 1 dereferenceable(1) [[TMP0]])
```
Perhaps use wildcards here to only look for something specific like 'private_alloca' here?   That way the test isn't affected if there are mangling changes.   (Same for the name values below.)

premanandrao commented on clang/test/SemaSYCL/builtin-alloca-errors-device.cpp:
```diff
@@ -0,0 +1,111 @@
+#include <stddef.h>
+
+#include "Inputs/sycl.hpp"
+
```
Please add a comment here about what this test tests for.

premanandrao commented on clang/test/SemaSYCL/builtin-alloca-errors-host.cpp:
```diff
@@ -0,0 +1,18 @@
+// RUN: %clang_cc1 -fsyntax-only -fsycl-is-host -triple x86_64-unknown-unknown -verify -Wpedantic -fcxx-exceptions %s
```
Are ``-Wpedantic`` and ``-fcxx-exceptions`` necessary?

premanandrao commented on clang/test/SemaSYCL/builtin-alloca-errors-device.cpp:
```diff
@@ -0,0 +1,111 @@
+// RUN: %clang_cc1 -fsyntax-only -fsycl-is-device -triple spir64-unknown-unknown -verify -Wpedantic %s
```
Is ``-Wpedantic`` necessary?

premanandrao reviewed (commented):
CFE changes look very good to me.   Thank you!

Just some nits.

victor-eds commented:
@premanandrao I addressed your comments. I also removed some redundant tests:
- `SemaSYCL/builtin-alloca.cpp`: I was checking the host took the host definition. That does not test our new builtin, but preprocessor `SYCL_DEVICE_ONLY` macros. Dropped.
- `CodeGenSYCL/builtin-alloca.cpp`: Dropped host test for same reasons.

premanandrao reviewed (approved):
FE changes LGTM.
